ocaml/typing/typedecl.ml

```diff
@@ -1757,7 +1757,15 @@ let transl_type_decl env rec_flag sdecl_list =
      jkind checks *)
   List.iter (fun (checks,loc) ->
     List.iter (fun (ty,jkind) ->
-      match Ctype.constrain_type_jkind new_env ty jkind with
+      (* The use [check_type_jkind] rather than [constrain_type_jkind] here is
+         conservative. It ensures that the delayed checks don't succeed by
+         mutating type variables from the [temp_env] in a way that won't be
+         reflected in the final type decls and may be incompatible with them.
+         An alternative would be to beef up [check_constraints] and really make
+         sure we re-check any kind constraint that could arise from translating
+         the typedecl RHSs. See Test 41 in [tests/typing-layouts/basics.ml] for
+         a subtle example. *)
+      match Ctype.check_type_jkind new_env ty jkind with
```

Comment:
To improve error messages, it might be better to use `constraint_type_jkind` here but then check whether any mutation happened. That shouldn't be too hard: look at `ty` before the call. If it's a variable, remember its kind. Then check again after the call. That way, we can report a message that complains about how our inference algorithm is not really as clever as it could be.

Reply:
I think this is not so simple.  We might need to expand `ty` and deal with `@@unboxed` types in the mutually recursive group to see it is a variable.  And doing that expansion properly in the middle of recursive typedecl checking is just why `constrain_type_jkind` is so complicated.

So I did it a different way: if `check_type_jkind` fails, I try `constrain_type_jkind`.  If that succeeds, we know we're in the case where this PR changes the behavior. I print an error similar to the one issued for the other case in typedecl where the compiler is insufficiently clever.  I'm not sure this relatively rare case justifies the added complexity in typedecl, and as noted above this is sometimes not only the compiler being stupid, but see what you think.  (I'm not worried about the cost of checking this twice since it only happens on the error path)

Reply:
Yes, good points about the difficulty. Happy with your approach.